### ISSUE TITLE
PROPOSAL configurable down-scaling behaviour in ReplicaSets & Deployments

### ISSUE THREAD

**Issue by thesuperzapper** [kubernetes/kubernetes#107598]
## Background

The idea of letting users customize the way Deployments (ReplicaSets) remove Pods when `replicas` are decreased has been floating around [since at least 2017](https://github.com/kubernetes/kubernetes/issues/45509), with [other issues dating back to 2015](https://github.com/kubernetes/kubernetes/issues/4301).

As of [Kubernetes 1.22](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#api-change-8), the `controller.kubernetes.io/pod-deletion-cost` annotation proposed in [KEP-2255](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2255-pod-cost) is available in BETA.

While KEP-2255 is a good start, its usefulness is limited because the `pod-deletion-cost` annotation must be updated BEFORE the `replicas` count is decreased. This means existing Kubernetes resources like [HorizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) can't be used, as doing so would require creating a system to keep the `pod-deletion-cost` annotation up to date (which is a [stated anti-pattern](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost), as this would overload the apiserver with PATCH requests). Thus, to use KEP-2255 with autoscaling, you need to create your own autoscaler that updates the `pod-deletion-cost` annotations before patching the replicas down, which makes KEP-2255 inaccessible to all but the most advanced Kubernetes users.

This proposal is about extending KEP-2255 in a generic way, allowing users to customize how ReplicaSets remove Pods on scale-down.

_PS: If the proposal is hard to understand, I recommend reading the "examples" section first._

## Design Details

The general idea is to add a `scaleConfig` field to [`ReplicaSetSpec`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#replicasetspec-v1-apps) that defines a sequence of __sorting methods__ for deciding which Pods to kill if `replicas` is decreased. The type of this `scaleConfig` field would be an array of `ScaleConfig` (because the order matters).

__These `ScaleConfigs` can be viewed as an API with the following which takes INPUTS and returns OUTPUTS:__

- INPUTS:
  - `numberRequested` - int
     - The minimum total number of Pods to return
     - (Think of this as the number of Pods we need to scale down)
  - `podList` - array[string]
     - The list of Pod-IDs which this ScaleConfig will choose from
     - (Think of this as the current `undecidedPods`)
- OUTPUTS:
  - `chosenPods` - array[string]
    - The list of Pod-IDs which this ScaleConfig has chosen to be scaled down.
    - (Think of this as the Pods which can be killed immediately)
  - `undecidedPods` - array[string]
    - The list of Pod-IDs which this ScaleConfig can't decide between.
    - (Think of this as the Pods which were "tied", which we will need to kill 1 or more of, but not all)
- NOTES:
   - If there are no "ties", then `len(chosenPods) == numberRequested` and `len(undecidedPods) == 0` 
   - If there are "ties", then `len(chosenPods) < numberRequested` and `len(undecidedPods) >= 2`

__The scale-down process for the ReplicaSet would be updated to use these `ScaleConfigs`:__

1. Kill Pods in `Pending` or `Unschedulable` states
2. Use the new `scaleConfigs` array to choose which Pods to kill first:
    1. Add a hidden "random choice" ScaleConfig to the end of the `scaleConfigs` array
        - (The "random choice" scaleConfig will always return `numberRequested` Pods from the `podList` sent to it)
    2. Let `numberRequested` be the number of Pods we need to kill
    3. Let `podList` be all the `Running` Pods
    4. For each ScaleConfig in `scaleConfigs`:
        1. Start a timeout counter for this ScaleConfig
            - (If the timeout counter has reaches `0`, skip to the next ScaleConfig)
        2. Let `chosenPods` and `undecidedPods` be the outputs of sending `numberRequested` and `podList` to the current ScaleConfig
        3. For each Pod in `chosenPods`:
            - Kill the Pod
            - Decrement `numberRequested` by `1`
        4. Replace the values of `podList` with those from `undecidedPods`

___NOTE:__ the current process is [described here](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#scaling-a-replicaset), and coded in [controller_utils.go#L753-L884](https://github.com/kubernetes/kubernetes/blob/14982cf5a94404118dce31c5f30b1578011eb438/pkg/controller/controller_utils.go#L753-L884)_

## Examples

Because this might be hard to visualize, I have written out some examples which show this feature being used in a number of situations. (Please refer to later sections for details on how each `ScaleConfig` is implemented, and its possible configs.)

### EXAMPLE 1: Current Defaults

This example shows the `scaleConfig` which corresponds to the current default `ReplicaSet` behaviour ([described here](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#scaling-a-replicaset)).

This will be the "implied default" when `scaleConfig` is not specified in a `ReplicaSet` to preserve backward compatibility.

```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: my-replicaset
spec:
  replicas: 3
  selector:
    matchLabels:
      app: my-app
  template:
    metadata:
      labels:
        app: my-app
    spec:
      containers:
        - name: my-container
          image: my-image

  ## this is the new section ##
  scaleConfig:
    - costAnnotationScaleConfig:
        annotationKey: "controller.kubernetes.io/pod-deletion-cost"
    - nodeReplicasScaleConfig:
        descending: true
    - podAgeScaleConfig:
        descending: true
```

### EXAMPLE 2: Random

This example shows a `scaleConfig` which uses __random selection__ to determine which Pods to kill.

This is to say, if you explicitly specify an empty list `[]`, Pods will be chosen at random when scaling down `replicas`.
This is a bi-product of the implied random choice as the last ScaleConfig, which is in place to deal with ties.

```yaml
apiVersion: apps/v1
kind: ReplicaSet
...
spec:
  ## this is the new section ##
  scaleConfig: []
```

### EXAMPLE 3: API

This example shows a `scaleConfig` which uses a central HTTP API to determine which Pods to kill.

The API contract is identical to the described "INPUTS" and "OUTPUTS" above, with payloads being encoded as JSON objects. (IN: `numberRequested`, `podList` OUT: `chosenPods`, `undecidedPods`)

```yaml
apiVersion: apps/v1
kind: ReplicaSet
...
spec:
  ## this is the new section ##
  scaleConfig:
    - apiScaleConfig:
        http:
          host: my-app-backend.my-namespace.svc.cluster.local
          httpHeaders: {}
          path: "/pod-deletion-cost"
          port: 80
          scheme: "http"
        maxRetries: 3
        timeoutSeconds: 5
```

### EXAMPLE 4: Heuristic Cost Probe

This example shows a `scaleConfig` which uses an `exec` probe against the Pods themselves to determine which Pods are least expensive to kill.

This is a heuristic approach because it's not scalable to always check every Pod before scaling down.

In this case, we are probing the first `sqrt(num_pods) + 1` Pods, in descending order of their Pod name, and would kill the Pod which returns the lowest cost from this sample.

```yaml
apiVersion: apps/v1
kind: ReplicaSet
...
spec:
  ## this is the new section ##
  scaleConfig:
    - heuristicCostProbeScaleConfig:
        probe:
          containerName: "my-container"
          exec:
            command:
              - /bin/my-app
              - pod-deletion-cost
          maxRetries: 3
          timeoutSeconds: 1
        sampleMethod: "name_descending"
        sampleSizeMethod: "sqrt+1"
        minSampleSize: 1
```

## ScaleConfig Types

This section lists and describes the different types of ScaleConfigs in detail.

_I expect this list will be refined over time, but this is a good start that I think covers most use-cases._

### ROOT: ScaleConfig

The root `ScaleConfig` type is similar to the current [`Volume v1 core`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#volume-v1-core) API.

__YAML Fields -- `ScaleConfig`:__

| Field                           | Type                          | Description                                                                                                     |
|---------------------------------|-------------------------------|-----------------------------------------------------------------------------------------------------------------|
| `timeoutSeconds`                | int                           | Timeout for a ScaleConfig before skipping to the next ScaleConfig. (default: `60`)                              |
| `costAnnotationScaleConfig`     | CostAnnotationScaleConfig     | Reads a Pod annotation containing "pod deletion cost" (similar to `controller.kubernetes.io/pod-deletion-cost`) |
| `apiScaleConfig`                | APIScaleConfig                | Asks a central API to return `chosenPods` and `undecidedPods` when passed `numberRequested` and `podList`       |
| `heuristicCostProbeScaleConfig` | HeuristicCostProbeScaleConfig | Asks a sample of Pods to report their "pod deletion cost" through `http`/`exec` probes                          |
| `nodeReplicasScaleConfig`       | NodeReplicasScaleConfig       | Reads the number of replicas on each Node                                                                       |
| `podAgeScaleConfig`             | PodAgeScaleConfig             | Reads the creation time of each Pod                                                                             |

### TYPE 1: CostAnnotationScaleConfig

The `CostAnnotationScaleConfig` works in much the same way as the KEP-2255 [`controller.kubernetes.io/pod-deletion-cost`](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) Pod annotation.

__Pseudocode:__

1. (INPUTS: `numberRequested`, `podList`)
2. Let `sortedPodsWithMetric` be an EMPTY sorted-array of `(podId, podDeletionCost)` tuples, sorted by `podDeletionCost` in ascending order
3. For each Pod in `podList`:
    1. Let `podDeletionCost` be the value of the annotation with `annotationKey` key:
        - (Default to `0` if Pod has no annotation with `annotationKey` key)
    2. Insert the `(podId, podDeletionCost)` tuple into `sortedPodsWithMetric` at the correct sorted position
4. Construct `chosenPods` and `undecidedPods` from the `sortedPodsWithMetric` array:
     - (Use the process described in "APPENDIX 1")
5. (OUTPUTS: `chosenPods`,  and an empty `undecidedPods`)

__YAML Fields -- `CostAnnotationScaleConfig`:__

| Field           | Type   | Description                                                                       |
|-----------------|--------|-----------------------------------------------------------------------------------|
| `annotationKey` | string | The annotation key to use (default: `controller.kubernetes.io/pod-deletion-cost`) |

### TYPE 2: APIScaleConfig

The `APIScaleConfig` allows users to specify an external HTTP API that returns `chosenPods` and `undecidedPods` when passed `numberRequested` and `podList`.

This is useful for scenarios when there is some central system that knows the cost of killing each Pod at all times.
For example in many systems that have a "master" and "worker" concept, the master will know the cost of killing each worker.

__Pseudocode:__

1. (INPUTS: `numberRequested`, `podList`)
2. Let `jsonPayload` be a JSON object with the following fields:
    - `numberRequested` (`int`)
    - `podList` (`array[string]`)
3. Let `jsonResponse` be the result of sending an HTTP POST with `jsonPayload` to the API, having the following fields:
    - `chosenPods` (`array[string]`)
    - `undecidedPods` (`array[string]`)
4. (OUTPUTS: `chosenPods`,  `undecidedPods`)

__YAML Fields -- `APIScaleConfig`:__

| Field            | Type         | Description                                                           |
|------------------|--------------|-----------------------------------------------------------------------|
| `http`           | HTTPScaleAPI | An HTTP API that returns the deletion cost of Pods in this ReplicaSet |
| `maxRetries`     | int          | The maximum number of retries (default: `3`)                          |
| `timeoutSeconds` | int          | The maximum number of seconds API call can run (default: `1`)         |

__YAML Fields -- `HTTPScaleAPI`:__

_NOTE: this is the SAME as [HTTPGetAction v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpgetaction-v1-core)_

| Field         | Type             | Description                                                                                    |
|---------------|------------------|------------------------------------------------------------------------------------------------|
| `host`        | string           | Host name to connect to.                                                                       |
| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                          |
| `path`        | string           | Path to access on the HTTP server.                                                             |
| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. |
| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                 |

__YAML Fields -- `HTTPHeader`:__

- [HTTPHeader v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)
  - ___NOTE:__ possibly needs to be extended with `valueFrom` to allow mounting headers from Secrets, for authentication_

### TYPE 3: HeuristicCostProbeScaleConfig

The `HeuristicCostProbeScaleConfig` is a heuristic (i.e. not 100% accurate) method which uses Pod probes to determine the cost of each Pod.

The general idea is to probe a sample of the Pods (but not all) for their `podDeletionCost` using an `exec` or `http` probe, and return the lowest cost Pods found in the sample.

The returned cost will be one of:
- `cost = 0`:
   - (Pod is able to be killed immediately - allows us to end sample early and NOT check all Pods)
   - (NOTE: this `0` has a different meaning to a `0` in  `controller.kubernetes.io/pod-deletion-cost`)
- `cost = 1+`:
   - (Pod costs `n` much to be killed - assume that all Pods return comparable values - kill lowest)

The sample can be made "smarter" by making it non-random. For example, we allow sampling the Pods with the lowest alphanumeric names first. (Allows applications to reduce the impact of a scale-down by preferring to assign workloads to Pods with the lowest name)

__Pseudocode:__

1. (INPUTS: `numberRequested`, `podList`)
2. Let `sampleSize` be the MAXIMUM of `minSampleSize` and the value calculated from `sampleSizeMethod`:
    - `sqrt+1`: sample `sqrt(len(podList)) + 1` Pods
    - `XX%`: sample `XX%` of the Pods in `podList`
    - `all`: sample all Pods in `podList`
3. Let `samplePods` be the first `sampleSize` elements of `podsList`, sorted by `sampleMethod`:
    - `random`: Randomly sort the Pods
    - `name_acending`: Sort the Pods by their name in ascending order
    - `name_decending`: Sort the Pods by their name in descending order
    - `age_acending`: Sort the Pods by their age in ascending order
    - `age_decending`: Sort the Pods by their age in descending order
4. Let `zeroCostCount` be an integer with initial value `0`
5. Let `sortedPodsWithMetric` be an EMPTY sorted-array of `(podId, podDeletionCost)` tuples, sorted by `podDeletionCost` in ascending order
6. For each Pod in `samplePods`:
    1. Start a timer of `timeoutSeconds` seconds:
        - (If this timer hits `0`, insert a `(podId, +inf)` tuple into `sortedPodsWithMetric`, and continue to next loop)
    2. If the probe is `exec` type:
        1. Run the `command` inside `containerName` on the Pod:
            - The final line of STDOUT should contain the `podDeletionCost` of the Pod
            - If the command returns a non-zero exit code, or malformed STDOUT, retry the probe up to `maxRetries` times
            - (If retries are exhausted, insert a `(podId, +inf)` tuple into `sortedPodsWithMetric`, and continue to next loop)
        2. Insert the `(podId, podDeletionCost)` tuple into `sortedPodsWithMetric` at the correct sorted position
        3. If `podDeletionCost = 0`:
            - Increment `zeroCostCount` by `1`
        4. If `zeroCostCount = numberRequested`, break the innermost for loop:
            - (Exits the sample early because we have found `numberRequested` 0-cost Pods)
    3. If the probe is `http` type:
        - (Same as for `exec` but using an HTTP GET instead of a command)
7. Construct `chosenPods` and `undecidedPods` from the `sortedPodsWithMetric` array:
    - (Use the process described in "APPENDIX 1")
8. (OUTPUTS: `chosenPods`, `undecidedPods`)

__YAML Fields -- `HeuristicCostProbeScaleConfig`:__

| Field              | Type      | Description                                                                                                                          |
|--------------------|-----------|--------------------------------------------------------------------------------------------------------------------------------------|
| `probe`            | CostProbe | The `exec`/`http` cost probe to use                                                                                                  |
| `sampleMethod`     | string    | The method used to sample Pods, `random`, `name_ascending`, `name_descending`, `age_ascending`, `age_descending` (default: `random`) |
| `sampleSizeMethod` | string    | The method used to determine the sample size `sqrt+1`, `XX%`, `all` (default: `sqrt+1`)                                              |                                                                                                                                   |
| `minSampleSize`    | int       | The minimum number of Pods in any sample (default: `1`)                                                                              |

__YAML Fields -- `CostProbe`:__

| Field            | Type              | Description                                                                                                    |
|------------------|-------------------|----------------------------------------------------------------------------------------------------------------|
| `containerName`  | string            | The name of a container in the Pod run the command inside. If "", the first container is used. (default: `""`) |
| `exec`           | ExecCostAction    | The `exec` cost probe to use. Must be `null` if `httpGet` is set.                                              |
| `httpGet`        | HTTPGetCostAction | The `http` cost probe to use. Must be `null` if `exec` is set.                                                 |
| `maxRetries`     | int               | The maximum number of retries (default: `3`)                                                                   |
| `timeoutSeconds` | int               | The maximum number of seconds the probe can run before the Pod is skipped (default: `1`)                       |

__YAML Fields -- `ExecCostAction`:__

_NOTE: this is the same as [ExecAction v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#execaction-v1-core)_

| Field           | Type         | Description                                                                                                    |
|-----------------|--------------|----------------------------------------------------------------------------------------------------------------|
| `command`       | string array | A command to run inside the container, must return the "cost" on its final STDOUT line                         |

__YAML Fields -- `HTTPGetCostAction`:__

_NOTE: this is the SAME as [HTTPGetAction v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpgetaction-v1-core)_

| Field         | Type             | Description                                                                                                                   |
|---------------|------------------|-------------------------------------------------------------------------------------------------------------------------------|
| `host`        | string           | Host name to connect to, defaults to the pod IP.                                                                              |
| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                                                         |
| `path`        | string           | Path to access on the HTTP server.                                                                                            |
| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. Name must be an IANA_SVC_NAME. |
| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                                                |

__YAML Fields -- `HTTPHeader`:__

- [HTTPHeader v1 core](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)

### TYPE 4: NodeReplicasScaleConfig

The `NodeReplicasScaleConfig` is similar to ["step 3" in the current ReplicaSet scaling behaviour](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#scaling-a-replicaset).

The idea is to remove the Pods which share a Node with the most other replicas, reducing the Pod concentration on each Node.

__Pseudocode:__

1. (INPUTS: `numberRequested`, `podList`)
2. Let `sortedPodsWithMetric` be an EMPTY sorted-array of `(podId, numberAdjacentReplicas)` tuples, sorted by `numberAdjacentReplicas` in descending order (if `decending = true`)
3. For each Pod in `podList`:
    1. Let `numberAdjacentReplicas` be the number of replicas (from this ReplicaSet) on the same Node as this Pod 
    2. Insert the `(podId, numberAdjacentReplicas)` tuple into `sortedPodsWithMetric` at the correct sorted position
4. Construct `chosenPods` and `undecidedPods` from the `sortedPodsWithMetric` array:
    - (Use the process described in "APPENDIX 1")
5. (OUTPUTS: `chosenPods`,  and an empty `undecidedPods`)

__YAML Fields -- `NodeReplicasScaleConfig`:__

| Field        | Type | Description                                                              |
|--------------|------|--------------------------------------------------------------------------|
| `descending` | bool | When `true`, Nodes with more Replicas are killed first (default: `true`) |

### TYPE 5: PodAgeScaleConfig

The `PodAgeScaleConfig` is similar to ["step 4" in the current ReplicaSet scaling behaviour](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#scaling-a-replicaset).

The idea is to remove Pods which were created more recently, leaving the longer running Pods.

__Pseudocode:__

1. (INPUTS: `numberRequested`, `podList`)
2. Let `sortedPodsWithMetric` be an EMPTY sorted-array of `(podId, podCreationEpoch)` tuples, sorted by `podCreationEpoch` in descending order (if `decending = true`)
3. For each Pod in `podList`:
    1. Let `podCreationEpoch` be the creation time of the Pod as a Unix Epoch
    2. Insert the `(podId, podCreationEpoch)` tuple into `sortedPodsWithMetric` at the correct sorted position
4. Construct `chosenPods` and `undecidedPods` from the `sortedPodsWithMetric` array:
    - (Use the process described in "APPENDIX 1")
5. (OUTPUTS: `chosenPods`,  and an empty `undecidedPods`)

__YAML Fields -- `PodAgeScaleConfig`:__

| Field        | Type | Description                                                                          |
|--------------|------|--------------------------------------------------------------------------------------|
| `descending` | bool | When `true`, Pods that were created more recently are killed first (default: `true`) |

### APPENDIX 1: construct `chosenPods` and `undecidedPods` from `sortedPodsWithMetric`

__Description:__

Most `ScaleConfigs` will internally create a list of Pods that is sorted by some metric and then need to return the requested number of Pods, prioritized by this list.

Naturally, if there are ties in the metric, all the tied Pods are returned as "undecided", so the tie can be broken by the next ScaleConfig.

Note, it's possible to return "chosen" and "undecided" Pods at the same time. This happens if there are some non-tied Pods at the beginning of the list, but not enough to fully fulfill the requested number of Pods.

__This process can be viewed as function with signature:__

- INPUTS: 
   - `numberRequested`:
      - The minimum number of Pods to return.
   - `sortedPodsWithMetric`:
      - A list of `(podId, podMetric)`, sorted by `podMetric`
- OUTPUTS:
  - `chosenPods`:
     - The list of `podId` which unambiguously should be included
  - `undecidedPods`:
     - The list of `podId` which have tied `podMetric`, but are required to fulfill `numberRequested`

__Pseudocode:__

1. (INPUTS: `numberRequested`, `sortedPodsWithMetric`)
2. Let `chosenPods` be an EMPTY array of Pods
3. Let `undecidedPods` be an EMPTY array of Pods
4. Let `totalFound` be an integer with initial value `0`
5. Let `sliceStartIndex` be an integer with initial value `0`
6. While `totalFound` < `numberRequested`:
    1. Let `sliceEndIndex` be the value of `sliceStartIndex`
    2. Let `sliceEndMetric` be the `podMetric` of the Pod at index `sliceEndIndex` of `sortedPodsWithMetric`
    3. While `(sliceEndIndex + 1) < len(sortedPodsWithMetric)`:
        1. Let `nextMetric` be the `podMetric` of the Pod at index `sliceEndIndex + 1` of `sortedPodsWithMetric`
        2. If `sliceEndMetric != nextMetric`:
            - Break the innermost while loop
        3. Increment `sliceEndIndex` by `1`
    4. Let `numPodsInSlice` be `sliceEndIndex - sliceStartIndex + 1`
    5. If `numPodsInSlice > (numberRequested - totalFound)`:
        - (There are too many Pods which all have tied metric, so we can't decide between them)
        - Append the Pods from index `sliceStartIndex` until `sliceEndIndex` (inclusive) in `sortedPodsWithMetric` to `undecidedPods`
    6. Else:
        - (We have found less than or equal to the number of Pods we need)
        - Append the Pods from index `sliceStartIndex` until `sliceEndIndex` (inclusive) in `sortedPodsWithMetric` to `chosenPods`
    7. Increment `totalFound` by `numPodsInSlice`
    8. Replace the value of `sliceStartIndex` with `sliceEndIndex + 1`
7. (OUTPUTS: `chosenPods`, `undecidedPods`)

__Examples:__

|           | numberRequested | sortedPodsWithMetric                                 | chosenPods   | undecidedPods          |
|-----------|-----------------|------------------------------------------------------|--------------|------------------------|
| EXAMPLE_1 | `2`             | `[("a", 0), ("b", 1), ("c", 3), ("d", 4)]`           | `["a", "b"]` | `[]`                   |
| EXAMPLE_2 | `2`             | `[("a", 0), ("b", 0)]`                               | `["a", "b"]` | `[]`                   |
| EXAMPLE_3 | `2`             | `[("a", 1), ("b", 1), ("c", 2), ("d", 3)]`           | `["a", "b"]` | `[]`                   |
| EXAMPLE_4 | `2`             | `[("a", 1), ("b", 1), ("c", 1), ("d", 2), ("e", 3)]` | `[]`         | `["a", "b", "b"]`      |
| EXAMPLE_5 | `2`             | `[("a", 0), ("b", 1), ("c", 1), ("d", 10)]`          | `["a"]`      | `["b", "c"]`           |
| EXAMPLE_6 | `3`             | `[("a", 1), ("b", 1), ("c", 2), ("d", 2), ("e", 2)]` | `["a", "b"]` | `["c", "d", "e"]`      |
| EXAMPLE_7 | `2`             | `[("a", 1), ("b", 1), ("c", 1), ("d", 1)]`           | `[]`         | `["a", "b", "c", "d"]` |
| EXAMPLE_8 | `1`             | `[("a", 1)]`                                         | `["a"]`      | `[]`                   |

__Python Implementation:__

```python
from typing import List, Tuple

def choose_pods(number_requested: int, sorted_pods_with_metric: List[Tuple[str, int]]) -> (List[str], List[str]):
    chosen_pods = []
    undecided_pods = []

    total_found = 0
    slice_start_index = 0

    while total_found < number_requested:
        slice_end_index = slice_start_index
        slice_end_metric = sorted_pods_with_metric[slice_end_index][1]

        while (slice_end_index + 1) < len(sorted_pods_with_metric):
            next_metric = sorted_pods_with_metric[slice_end_index + 1][1]
            if slice_end_metric != next_metric:
                break
            slice_end_index += 1

        nun_pods_in_slice = slice_end_index - slice_start_index + 1
        if nun_pods_in_slice > (number_requested - total_found):
            undecided_pods += [pod_tuple[0] for pod_tuple in
                               sorted_pods_with_metric[slice_start_index:slice_end_index + 1]]
        else:
            chosen_pods += [pod_tuple[0] for pod_tuple in
                            sorted_pods_with_metric[slice_start_index:slice_end_index + 1]]

        total_found += nun_pods_in_slice
        slice_start_index = slice_end_index + 1

    return chosen_pods, undecided_pods

def example_1():
    number_requested = 2
    sorted_pods_with_metric=[("pod_a", 0), ("pod_b", 1), ("pod_c", 3), ("pod_d", 4)]
    
    chosen_pods, undecided_pods = choose_pods(number_requested, sorted_pods_with_metric)
    
    assert chosen_pods == ["pod_a", "pod_b"]
    assert undecided_pods == []
    
    print("EXAMPLE 1:")
    print(f"INPUTS:  number_requested={number_requested}, sorted_pods_with_metric={sorted_pods_with_metric}")
    print(f"OUTPUTS: chosen_pods={chosen_pods}, undecided_pods={undecided_pods}")   
    
example_1()
```

**Comment by alculquicondor** [comment]
FYI @gaocegege

**Comment by alculquicondor** [comment]
Technical nits:
- I didn't understand what "a chosen pod" is. Chosen for deletion? But then what is the difference with "pods to kill".
- Currently the ages are not directly compared. We compare the logarithm of them (truncated) to allow for some randomness. Maybe we could disable the logarithm calculation in the API, or even parameterize the base. But for simplicity I would leave it always enabled. The name `podAgeScaleConfig` would have to change to reflect that.

Other than that, the proposal is very interesting and I like the flexibility of it.
There are two innovations here: CostAPIScaleConfig and HeuristicCostProbeScaleConfig

Do you know of real world scenarios were you absolutely need both? I'm thinking that CostAPIScaleConfig should be enough and HeuristicCostProbeScaleConfig is too costly to pursue. Perhaps we could start with one and consider adding the other if there is demand.

**Comment by tbonfort** [comment]
> Do you know of real world scenarios were you absolutely need both? I'm thinking that CostAPIScaleConfig should be enough and HeuristicCostProbeScaleConfig is too costly to pursue. Perhaps we could start with one and consider adding the other if there is demand.

@alculquicondor we are currently using a custom autoscaler that is very similar to HeuristicCostProbeScaleConfig (with an http probe and sampleSizeMethod=all), not including it would be disappointing.

**Comment by alculquicondor** [comment]
But you could have your own controller running such probe, then RS controller only queries this controller through CostAPIScaleConfig. That gives you the same behavior without compromising the latency of the RS controller.

**Comment by tbonfort** [comment]
> But you could have your own controller running such probe, then RS controller only queries this controller through CostAPIScaleConfig. That gives you the same behavior without compromising the latency of the RS controller.

It's an additional controller that needs to be developed, maintained and deployed. That additional controller would still need to query all pods at scale down time, thus with the same latency as if the RS controller were querying them directly (unless it is acceptable for that custom controller to provide stale/cached metrics)

**Comment by alculquicondor** [comment]
> unless it is acceptable for that custom controller to provide stale/cached metrics

I think it should be. It's a heuristic anyways. And you can't realistically assume that, by the time the RS controller starts removing the pods, the metrics didn't change.

> It's an additional controller that needs to be developed, maintained and deployed.

I think that's acceptable to avoid compromising the throughput of the RS controller.

In any case, this is just my POV, and I'm only a reviewer in SIG Apps. Also, from my POV, this proposal looks ready for a KEP and a Production Readiness Review. But maybe it's better if you bring the proposal to the next SIG meeting before putting up a KEP.

**Comment by thesuperzapper** [comment]
@alculquicondor @tbonfort thanks for the feedback!

I have made some significant updates to the proposal based on your comments:

1. The "pods to kill" vs  "chosen pods" was ill-defined, so I have removed it and simplified the OUTPUTS of a ScaleConfig to:
    - `chosenPods`: the list of Pod-IDs which this ScaleConfig has chosen to be scaled down.
    - `undecidedPods`: the list of Pod-IDs which this ScaleConfig can't decide between.
2. After the change in "1", I realized that all ScaleConfigs will likely use the same function internally when populating `chosenPods` and `undecidedPods`. I have updated all ScaleConfigs to use a common function described in "APPENDIX 1".
3. I realized that `CostAPIScaleConfig` had nothing to do with cost, so I have renamed it to `APIScaleConfig`. (It's just an external implementation of the `ScaleConfig` API)

@alculquicondor I would greatly appreciate another look over if you have time

**Comment by thesuperzapper** [comment]
@alculquicondor  I am happy to present this proposal at the next Apps SIG meeting (but it would be at 4am my time, Australia).

How would I go about doing this?

**Comment by alculquicondor** [comment]
You add it to the agenda and a lead will approve it if there is enough space left in the meeting.

**Comment by thesuperzapper** [comment]
@alculquicondor I have done the following:
- Sent an email to https://groups.google.com/g/kubernetes-sig-apps (but it has not shown up)
- Added an agenda item to the SIG Apps meeting notes.

**Comment by soltysh** [comment]
Have you seen https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2185-random-pod-select-on-replicaset-downscale which covers your random example?

**Comment by thesuperzapper** [comment]
@alculquicondor @tbonfort @soltysh (and others watching)

__We discussed this proposal in the SIG Apps meeting today, the main points of discussion were:__
- there are problems with the `pod-deletion-cost` annotation from KEP-2255:
    1. it can't be used with existing scaling tools like `HorizontalPodAutoscaler` (annotations must be set BEFORE scale)
    2. even a custom ReplicaSet "scaler controller" may still create too many PATCH requests (overloading the API server)
- implementing the `HeuristicCostProbeScaleConfig` might be complex:
   - the ReplicaSet controller cant directly interact with Pods to perform `exec` or `http` requests (only the kubelet can)
   - possible workaround (might be too complex):
      1. the ReplicaSet controller chooses the Pods in the sample
      2. the ReplicaSet controller "marks" these Pods to be probed
      3. the kubelet probes the Pods and creates a Pod-event with the cost information 
          - (events are inherently temporary, so are better than annotation/status fields, which would become out of date if not cleared)
      4. the ReplicaSet controller waits for these events and uses them to make the scale-down decision
- the `APIScaleConfig` is the most powerful part of this proposal:
   - the user can implement their "scale API" using any approach they would like
   - it is better than KEP-2255 because it's PART of the scale-down flow (when `replicas` is decreased), rather than requiring the annotations to already be set BEFORE scale-down starts

---

__As a result, I plan to create a KEP with ONLY the `APIScaleConfig` from this proposal:__
- I now realize that `APIScaleConfig` can be used to implement the other ScaleConfigs in this proposal:
    - e.g. `HeuristicCostProbeScaleConfig` can be implemented by having the "scale API" run bash/http probes on the Pods (which the ReplicaSet controller cant do)
    - e.g. `PodAgeScaleConfig` can be implemented by giving the "scale API" access to read Pod information
- I will propose that an `apiScaleConfig` field (or similar name) be added to the root of `ReplicaSetSpec`/`DeploymentSpec`:
   - if `apiScaleConfig` is unset, default to current behavior
   - if `apiScaleConfig` is set, first use the "scale API" to choose Pods to kill, if there are any "undecided" Pods, use the current behavior to decide between the "undecided" Pods
- I think we should create a sample/reference "scale API" for people to use:
   - it could even use the `scaleConfig` array from the original proposal as its configuration

**Comment by thesuperzapper** [comment]
@alculquicondor @soltysh I have now raised a KEP https://github.com/kubernetes/enhancements/issues/3189, and associated initial PR of the KEP summary: https://github.com/kubernetes/enhancements/pull/3190

I would greatly appreciate your review.

**Comment by alculquicondor** [comment]
Just FYI, enhancements freeze is next Thursday, so it might be too late to get this merged. I'll review ASAP anyways.

**Comment by thesuperzapper** [comment]
For those watching, I have raised a proposal for the "Downscale Pod Picker" API to be implemented in the [OpenKruise](https://openkruise.io/) project in the same way as [KEP-3189](https://github.com/kubernetes/enhancements/issues/3189). Hopefully, this will get this sorely needed feature into user's hands sooner.

See the proposal here: https://github.com/openkruise/kruise/issues/902

**Comment by thesuperzapper** [comment]
After working on this problem for a while, I think there is a better way to implement selective down-scaling than what I proposed in [KEP-3189](https://github.com/kubernetes/enhancements/issues/3189). 

We already have the special `/apis/apps/v1/namespaces/{namespace}/deployments/{name}/scale` API that allows autoscalers (like HPA) to get/patch the "replica count" and "label-selector" in a resource-agnostic way, we can extend this API to include a selective-downscale feature. 

_NOTE: the REST payload for the `/scale` API is the [`Scale v1 autoscaling`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#scale-v1-autoscaling) object._

## Possible Implementation

We could extend `Scale v1 autoscaling` with a function to temporarily set the `pod-deletion-cost` of one-or-more Pod.

This would mean that the [`controller.kubernetes.io/pod-deletion-cost`](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) annotations don't have to be set before initiating the scale (thus also don't have to be cleaned up if K8S decided to not remove the Pod)!

For example, here is `Scale v1 autoscaling` with an added `spec.podDeletionCosts` field:

```yaml
apiVersion: autoscaling/v1
kind: Scale
metadata: ...

# NOTE: `spec` is used with PATCH requests to `/scale` subresource to initiate a down/up scale
spec:
  replicas: 10

  # \/ new field \/
  # overrides the pod-deletion-cost of the listed pods (but only for the duration of this scale)
  podDeletionCosts:
    - name: "my-pod-xxxxx"
      cost: "-100"
    - name: "my-pod-yyyy"
      cost: "100"  
```

### Other thoughts:

- There is a question of how to get information about the temporary "pod-deletion-costs" to the ReplicaSet controller, one option is to store them in a JSON blob within an Event on the ReplicaSet.
- The `controller.kubernetes.io/pod-deletion-cost` feature is only for ReplicaSets, but we might want to support this "selective downscaling" in other resource types (possibly even CRDs), and to do this it might make more sense to remove the explicit association to "pod-deletion-cost" and instead provide a sorted list of "best pods to remove".
   - _We could extend the [`CustomResourceSubresourceScale v1`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#customresourcesubresourcescale-v1-apiextensions-k8s-io) to allow users to indicate how their controller wants to receive these "best pods to remove"._

---

/remove-lifecycle stale

**Comment by ddelange** [comment]
Do I understand correctly that in the end, an end-user could set up a HPA with something like a 'scale down most idle' behaviour, and HPA would set (instead of annotations) the appropriate pod deletion costs in the Scale spec? Would that be based on current CPU usages from the metrics server, and injected just before the scale down occurs?

**Comment by thesuperzapper** [comment]
> Do I understand correctly that in the end, an end-user could set up a HPA with something like a 'scale down most idle' behaviour, and HPA would set (instead of annotations) the appropriate pod deletion costs in the Scale spec? Would that be based on current CPU usages from the metrics server, and injected just before the scale down occurs?

@ddelange With the extension to the `/scale` API proposed in https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884, any autoscaler (not just HPA) will be able to influence which Pods are removed during a downscale. 

There would need to be a separate discussion (and probably sperate KEP) about what new options could be added to HPA that make use of the new `/scale` features, and something like 'scale down most idle' would certainly be a good contender.

To move this forwards, I would love to get some thoughts from @sftim @thockin @alculquicondor about if we should create a new KEP based on https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884.

**Comment by alculquicondor** [comment]
Can you expand on how would this fit the use cases? How would these podDeletionCosts field be populated?

**Comment by ahg-g** [comment]
I think populating the costs in this new proposal would be similar to how we assumed users/controllers would do in https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2255-pod-cost; but I thought one of the main pushbacks on the existing beta feature was exactly that, and the hope was to find a way for pod costs to be populated by the pods themselves?

@thesuperzapper perhaps it is worth having a quick comment comparing how this approach handles the drawbacks of the existing feature?

**Comment by thockin** [comment]
I like this approach much better than https://github.com/kubernetes/enhancements/pull/3190 - we should set a "reasonable" limit on how many pods you can put in here, like 10 or 100 or 256, but not unbounded and probably not 1000.  I do like how this lets you spec negative values or positive, so you can protect or sacrifice as needed, based on available information.

@abdullah, even without a way for pods to report their own costs, this seems useful - if the autoscalar has material information that the pod itself doesn't it can pass that here.  Would it make sense for these values to be additive to the existing pod-deletion-cost (if present)?  I still would like to see a way for pods to report their own cost, but I don't think that needs to block this, do you?  Asking, not telling :)

**Comment by ahg-g** [comment]
> @abdullah, even without a way for pods to report their own costs, this seems useful - if the autoscalar has material information that the pod itself doesn't it can pass that here. Would it make sense for these values to be additive to the existing pod-deletion-cost (if present)? 

This certainly is useful and offers a better API than the existing annotation-based feature, I think it should replace the existing one though and not get into the weeds of how they could co-exist. 

My only concern is how job-like APIs which may not have a scale subresource will benefit from this.

> I still would like to see a way for pods to report their own cost, but I don't think that needs to block this, do you? Asking, not telling :)

I agree. I asked about it mainly because it was one of the reasons for the push back on the existing pod-deletion-cost feature.

**Comment by exArax** [comment]
Is the method of pod-deletion-cost currently implemented in a specific version of Kubernetes? It would be really helpful to have a configuration that would enable the deletion of specific replicas.

**Comment by ddelange** [comment]
please see the first three paragraphs of the issue description. regarding the current implementation, please see [docs](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost)

**Comment by ddelange** [comment]
> the hope was to find a way for pod costs to be populated by the pods themselves

@ahg-g is it a feasible option to set PodDeletionCost on all pods before a scale down happens?

In the case of the 'scale down most idle' semantic, there needs to at least be knowledge about all pods. With that knowledge, you could add a negative PodDeletionCost for the X most idle pods, before initiating a scale down which will delete those X pods.

Then there's also no need to cleanup 'dangling' cost annotations (which the proposal of @thesuperzapper would also solve, by not setting them on the Pod at all https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884).

**Comment by thesuperzapper** [comment]
I still think we should aim to implement an extension to the `/scale` API that allows setting pod-deletion-cost, like i proposed in https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884.

 However, I would greatly appreciate someone else tacking up the work on this, as I really don't know when I will get time to implement it. 

___PS:__ if you are someone who is willing and able to push this forwards, I am happy to discuss the idea further with you, but will probably not have time to actually help with the code._

**Comment by nrjohnstone** [comment]
Most idle based on what? CPU? This still won't take into account jobs that are potentially doing long running processing that maybe in the middle of an await on some external IO ... All of the downscaling ideas seem to be biased towards "APIs" that usually only execute for short duration calls, and don't seem to take into account potentially how jobs are used... What is wrong with implementing an HTTP endpoint similar to the health checks and termination checks and returns a value... You could go down the "idle/working" route and return a true/false .. in which case while downscaling each of the pods are queried to find a pod that is "idle" and happy to be killed off.... If the downscaling still has to happen, you could have a second endpoint that returns the "cost" for stopping that pod... Then you could either get a % of the pods costs and downscale the lowest, or you could implement some kind of tiered system where the number has to be between 1 and 100 and any pod that returns <10 gets downscaled.... then if more still need to be downscaled, do another round and kill any pod < 20 etc... This leaves it up to the users how to calculate their "cost" ... Most of this depends on how "aggressive" you want the downscaling to happen I guess, and for our usecases we want to downscale but we are quite happy to wait 60 minutes until a pod finishes it's current job.

At the moment one of the ways we achieve this is using the pre-stop hook to hit an HTTP endpoint "/drain" that signals to the pod it should finish what it is doing and NOT dequeue any more messages then shutdown and have a wait of 60 minutes. This seems to work fine with the only annoying thing being that the pod state is "Terminating" for quite a long time...

**Comment by nrjohnstone** [comment]
> > the hope was to find a way for pod costs to be populated by the pods themselves
> 
> @ahg-g is it a feasible option to set PodDeletionCost on all pods before a scale down happens?
> 
> In the case of the 'scale down most idle' semantic, there needs to at least be knowledge about all pods. With that knowledge, you could add a negative PodDeletionCost for the X most idle pods, before initiating a scale down which will delete those X pods.
> 
> Then there's also no need to cleanup 'dangling' cost annotations (which the proposal of @thesuperzapper would also solve, by not setting them on the Pod at all [#107598 (comment)](https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884)).

The issue I have with these approaches is who is going to set the PodDeletionCost ? The only thing really that knows how expensive a pod is to delete is the pod itself... Which either means (1) the pod knows how to set Kubernetes labels, meaning that now you don't have an agnostic container that can run anywhere or (2) You have some kind of operator that can query all of the pods and set the deletion cost label on each one.... which begs the question how does it query the pod for it's cost ... which goes back to my other post above about having some specific HTTP endpoints that can be implemented to return this cost...

**Comment by thockin** [comment]
> The only thing really that knows how expensive a pod is to delete is the pod itself

I think there are cases where that is true, and cases where is is false.

Lots of discussion on https://github.com/kubernetes/enhancements/pull/3190

e.g. https://github.com/kubernetes/enhancements/pull/3190#pullrequestreview-868561920 was my initial reaction, much of which still seems appropriate to ask.

**Comment by thesuperzapper** [comment]
I agree with @thockin, there are _some situations_ where individual pod is the only thing which knows its "deletion cost" However, there are _almost no situations_ where the Pods know their "relative deletion costs" compared with all other Pods in the Deployment, and this is actually what we care about when deciding which Pods to remove in a scale down.

Therefore, if a user wants to autoscale, they will inevitably need to create an "autoscaler system" which takes signals from ALL Pods when deciding the two most important things:

1. Would I like to change the number of replicas (based on load factor targets, for example)
2. If I want to decrease the replicas, which ones are the best to remove (least disruptive to users, for example)

Deployments in Kubernetes today are completely missing the ability to do the second one, which is why I think we should [extend the `/scale` API](https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884); this approach lets an "autoscaler system" suggest to Kubernetes which Pods are best to remove (in the form of relative deletion costs) at the same instant it requests a lower number of replicas, rather than hoping the Pods have up-to-date annotations/labels/status.

**Comment by sftim** [comment]
How about a separate, optional API for scaling hints.
```yaml
---
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: whatever-1234
  labels:
    app: whatever
spec:
  replicas: 2000
  selector:
    matchLabels:
      podlabel: foo
  template:
    metadata:
      labels:
        app: whatever
        podlabel: foo
        otherpodlabel: bar
    spec:
      containers:
      - name: whatever
        image: container-images.example/something/whatever:42
```

and the hints:
```yaml
---
apiVersion: example.example.example/v1alpha1
kind: ScalingHint
metadata:
  name: whatever
spec:
  target:
    selector: # selects (in this case) ReplicaSets
              # ie, any ReplicaSet labelled app=whatever
      matchLabels:
        app: whatever
    group: "" # core 
    resource: replicasets
  podDeletionHints:
    - fieldSelector: # selects Pods
        matchLabels:
          "metadata.name": whatever-1234-456 # example, could work differently
                                             # prefer to remove this one named Pod
      costAdjustment: -100
```

**Comment by thesuperzapper** [comment]
This issue is critical for supporting many workloads on Kubernetes. 

My view is that the __least disruptive__ fix is extending the `/scale` API as I described in: https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1210060884

This approach extends the [existing `pod-deletion-cost` annotation](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost), by allowing "temporary" costs to be assigned to specific pods for the duration of a specific scale-down request.

_I have not got the cycles to pursue this right now, so if anyone else sees the value and has some time to work on it, please comment!_

---

/remove-lifecycle stale

**Comment by alculquicondor** [comment]
/sig autoscaling
/wg-batch

I'm not sure about adding `/help` here, given that this is not an easy problem to solve. `/help` might give the wrong expectations that this can be implemented right away.

**Comment by kerthcet** [comment]
This made me think of someone from the community asking for the kube-scheduler respecting the podTopologySpread when scaling down, maybe also a useful case for this topic.

**Comment by alculquicondor** [comment]
kube-scheduler is not involved in scale down. It's the replicaset controller.

**Comment by remiville** [comment]
Hi,

For a certain set of pods, I'm using the pod deletion cost annotation with a custom replica controller to choose which pods to delete when scaling down.

Now, for another type of pods for which I use a standard HPA, I have another concern about forbidding scale down to delete some critical pods doing something (e.g. "long running processing that maybe in the middle of an await on some external IO" as mentioned by @nrjohnstone).
I use the prestop hook to avoid these pods to be deleted while they're doing something critical, but this is not great, because if it is a long task, I think k8s removes the pod from the network since an undetermined amount of time, but my pod contains an application server which needs to often communicate with a database or return a response to another pod before quitting.

In this case the pod deletion cost is not sufficient because it only prioritizes candidates for deletion, but there I need to exclude some pods to be part of candidates.

There could be another annotation to achieve that.
Something better would be that the HPA object permits to define both a sort and filter function, the former to have the equivalent to the pod deletion cost annotation (but based on any label or probe), the latter to exclude some pods from deletion (again based on any label or probe).

Maybe you have another idea to achieve my goal.

**Comment by ddelange** [comment]
Depending on your use-case, I think with the [downward api](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/) and patch permission, it should be relatively straight forward to have your Pod patch it's own pod-deletion-cost to a high value when it's 'busy' (your definition here: crunching numbers, awaiting network requests, preparing a HTTP response, etc), and bring it back to 0 when it's 'idle' again.

Completely excluding a Pod from HPA scale down sounds like an anti-pattern to me. There is the `safe-to-evict` annotation which you could additionally (self)assign. And if your standard HPA is scaling down when it actually shouldn't (CPU-based metric but a non CPU-bound workload?) there is still custom scaling with e.g. keda.sh.

Related ref for CPU-based 'busy' switch: proposal to use the kserve sidecar container to monitor the CPU usage of the Pod: https://github.com/knative/serving/issues/13075#issuecomment-1556295956

Hope that helps :)

**Comment by remiville** [comment]
Thanks for your support @ddelange.

I don't think excluding pods from scale down is antipattern

**Comment by sftim** [comment]
> I don't think excluding some pods from scale down is anti pattern, for me it is a logical behaviour

It'd meet my understanding of an antipattern:
> a common response to a recurring problem that is usually ineffective and risks being highly counterproductive

(quoting Wikipedia)

> if at some point in time you don't want some pods to be scaled down (because they're doing something critical) you should be able to ask the HPA to not consider them as candidates for deletion

You can have a controller make two related assertions:
- there is a minimum scale for the relevant HorizontalPodAutoscaler
- the Pods that you'd like to protect have a higher deletion costs than their peers you are happy to replace

That can be implemented out-of-tree; for example:
- count the number of Pods targeted by an HPA that have some particular annotation
- ensure that the minimum scale for that HPA was at least that number of Pods
- set deletion costs for the targeted pods, so as to favor removal of the Pods lacking this particular annotation

Someone who wants to write that can write it today, so the obstacle is then more a lack of volunteers willing to make the controller I've sketched, rather than a shortcoming in Kubernetes' architecture and core API design.

---

What might be useful: a way to temporarily set a different minimum scale for an HPA, that is larger than the HPA's defined minimum scale in its `.spec`
If that mechanism existed, it would help someone making that out of tree controller. So, maybe they'd want to suggest that change to the HPA API?

**Comment by nrjohnstone** [comment]
Given that the code running inside the pod is the ONLY thing that knows it's true deletion cost e.g. Am I in the middle of an expensive operation and absolutely do not want to be shutdown, this now means one of the following

1. The code must now know about annotations and the Kubernetes API since it has to be able to apply annotations and remove them based on it's current execution context. I hope this is obvious to everyone why this is a huge anti pattern.
2. The code exposes an endpoint, similar to a health endpoint, that can be queried by the HPA to find out the deletion cost. This has issues with scale when running large numbers of pods and the HPA having to do a lot of querying.
3. Another option might be to expose an endpoint where a webhook can be registered that represents an abstraction that the service can call to register it's deletion cost at various points. This endpoint could be another service that is responsible for all of the annotation work so the HPA has recent information to work off.

The comment around relevant deletion cost makes little sense when you consider transactional workloads with long running processes.

If I am running 50 pods all doing "video conversion" (the classic example) and I've spent 40 minutes encoding/processing a stream, there is no way I want to be scaled down regardless of what the others are doing, I'd like to actually finish the processing, save the result THEN exit.

A lot of it depends what you are doing with the pods and why you are down scaling. In the situations where I want to scale from 0 to N for processing work items, usually I only want to scale back down once the work is complete. I don't just want to arbitrarily start scaling down while there is a ton of work remaining.

I could see situations however where you want the number of pods to start downscaling before the work was complete, again I've usually just handled this with my service code either exiting after waiting for work for a given idle time (10 seconds without any messages on a queue for example) which of course terminates the pod successfully.

If you wanted to have a "ratio" of pods based on, for example, the number of messages in a queue, then given the video processing example above, you could still advertise a "deletion cost" on an endpoint but something else would have to make the decision about which pods to kill off.

Taking the example further, let's say you wanted the deletion cost to simply be the % of the video stream that was processed already. This would allow all active pods to provide this (or push it to an external endpoint) and an orchestration process would have all the information to downscale the pods by killing off those that had done the least amount of work.

The vast majority of the implementations I've done have been scale from 0 to X, process all the work, then scale back to 0 by the worker pods shutting themselves down once they are idle for X seconds, and Keda achieves the scale up perfectly, and allowing the pods to terminate themselves means the HPA is not involved so you don't have to deal with annotations and all of that drama.

**Comment by nrjohnstone** [comment]
> Thanks for your support @ddelange.
> 
> I don't think excluding some pods from scale down is anti pattern, for me it is a logical behaviour: if at some point in time you don't want some pods to be scaled down (because they're doing something critical) you should be able to ask the HPA to not consider them as candidates for deletion.
> 
> I think using the prestop hook combined with a high grace period to delay the deletion of selected pods is more anti pattern: it gives the impression to the HPA the scale down is going well, because it sees some pods in terminating state (even if they still alive during a long time because of this hack), but in fact the scale down could have been achieved quickly by selecting other pods for deletion (those who are actually doing nothing critical from the application point of view).

The pre-stop hook was more of a hack due to the nature of the current HPA implementation... Definitely an anti-pattern for the exact reasons you mention

**Comment by sftim** [comment]
> Another option might be to expose an endpoint where a webhook can be registered that represents an abstraction that the service can call to register it's deletion cost at various points. This endpoint could be another service that is responsible for all of the annotation work so the HPA has recent information to work off.

Yes, and this could be a sidecar helper.

**Comment by sftim** [comment]
> Keda achieves the scale up perfectly

“Not using HorizontalPodAutoscaler” is already a supported pattern :wink:

**Comment by remiville** [comment]
> The vast majority of the implementations I've done have been scale from 0 to X, process all the work, then scale back to 0 by the worker pods shutting themselves down once they are idle for X seconds, and Keda achieves the scale up perfectly, and allowing the pods to terminate themselves means the HPA is not involved so you don't have to deal with annotations and all of that drama.

In my case I've not a big amount of pods and they are more or less application servers so they cannot shutdown themselves systematically when they're idle during a certain period of time.

I didn't know keda before, it is interesting.

What I would like to do is to let the hpa decide when scale out or in should be considered, classically based on pods cpu / memory usage.
But I would like to delegate the pod deletion step (e.g. to a custom controller pod) to choose the set of pods that can terminate themselves (without letting the replicaset recreating them automatically, like keda seems to achieve).

Maybe it is already possible to intercept scale in events, temporarily disable hpa/deployment, do the selective deletion, and finally reactivate hpa/deployment with an updated replica count.
Or maybe keda already manages this use case idk.

**Comment by alculquicondor** [comment]
For completion of the user story: who would write into these fields?

**Comment by remiville** [comment]
> After thinking about this a lot, I think we can significantly improve down-scaling (and help make stuff like KEDA better) by extending the [`/apis/apps/v1/namespaces/{namespace}/deployments/{name}/scale`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#scale-v1-autoscaling) API to optionally be more selective about which Pods are removed.
> 
> Specifically, we can allow the following additional fields beyond `spec.replicas`:
> 
> 1. `spec.temporaryDeletionCosts[]`
>       - a list of temporary [`pod-deletion-costs`](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) for just this scale 
>       - (stops the need to annotate the pods before initiating the scale-down, and to remove annotations after scaling)
> 1. `spec.priorityRemovalCandidates[]`
>       - a list of Pod ID/names that are more likely to be removed to meet the new `replicas` target, if they exist
>       - (allows the caller to influence which pods are removed, if they have some specific knowledge about which are the best candidates)

I'm ok with `spec.priorityRemovalCandidates[]`, but I don't understand the utility of `spec.temporaryDeletionCosts[]`.

**Comment by thesuperzapper** [comment]
> For completion of the user story: who would write into these fields?

@alculquicondor I relanyone who currently scales `replicas`.

This is really two groups:

1. __Automated Autoscalers:__
     - First-party autoscalers: HorizontalPodAutoscaler 
     - Third-party autoscalers: [KEDA](https://keda.sh/docs/2.13/concepts/scaling-deployments/)
     - Custom autoscalers: (could be as simple as a python script running a loop which knows about the app)
1. __Manually actions by admins:__
     - If the admin wants to downscale, but not kill a specific pod, they could use [`kubectl patch dep my-deployment --subresource='scale'  ...`](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#scale-kubectl-patch) and also set `temporaryDeletionCosts`.

**Comment by remiville** [comment]
> When you "terminate themselves" is this because you are wanting them to shutdown gracefully? This is already achieved if you handle a termination event properly inside your service or a pre-stop hook as well, but depending on how much work you have left to do (minutes versus seconds) this can be a bit of a hack as the pods show themselves "terminating" until it actually shuts down.

If I achieve to intercept the scale down and delegate it to a custom controller, I would like the latter to select idle pods to delete them, then the pre-stop hook is used only to correctly terminate opened I/O, not to finish a task.

> By application server I assume you mean these are just standard backend APIs that don't do long running requests?

It depends, here I have two types of pod, one server type, to save/retrieve data, one scheduler type which can delegate a task to another pod (created on demand) or execute the task directly.
Both types of pods can execute long running tasks from time to time (mostly the scheduler type, not really the server type but not impossible).

**Comment by alculquicondor** [comment]
> First-party autoscalers: HorizontalPodAutoscaler

I think you would have to complete the story about how HPA can know how to set those values.

**Comment by thesuperzapper** [comment]
After thinking about it even more, I have a more complete proposal than https://github.com/kubernetes/kubernetes/issues/107598#issuecomment-1965493996.

### Here is my new proposal: https://github.com/kubernetes/kubernetes/issues/123541

Please comment on that issue if you have specific feedback to the `Scale` subresource extension idea.

**Comment by remiville** [comment]
Do you think it is already possible to select pods to delete when a scale down is triggered by HPA using an admission controller (mutating web hook) ?

For example, pods that decide to be potential candidates for deletion update their pod-deletion-cost annotations to be prioritised for deletion, then send this  information to the admission controller, so the latter maintains a list of candidates for deletion.
when a scale down request is intercepted by the admission controller it checks if the new replica count is greater or equals to the number of candidates and changes the replica count accordingly (setting it to zero if no candidate).

**Comment by thesuperzapper** [comment]
> Do you think it is already possible to select pods to delete when a scale down is triggered by HPA using an admission controller (mutating web hook) ?

@remiville probably not, because you would also have to have a background process that removes any pod-deletion-cost annotations on pods which were not actually removed during the scale down.

This is of course unless you actually check every single pod's annotations each time any downscale is requested (which might not be possible if there is a large number of pods in the cluster).

**Comment by remiville** [comment]
@thesuperzapper ok so it may be possible if the webhook controller for example  uses the watch operation to know which pods are deletable or not (for example when when a pod is deletable it sets its pod-deletion-cost to 0, 1 otherwise).

I also considered the possibility of having pods with two labels deployment=app and scale=app with a Deployment targeting deployment=app pods and a HPA targeting scale=app pods.
If a pod must not be deleted by the HPA it removes its scale=app label.
But I think this latter solution has side effects, like triggering undesired scale outs for example.

**Comment by thesuperzapper** [comment]
Commenting to keep this issue alive.

To discuss my current proposal around extending the `Scale` sub resource, see:

https://github.com/kubernetes/kubernetes/issues/123541

**Comment by thesuperzapper** [comment]
Commenting to keep this issue alive.

To discuss my current proposal around extending the `Scale` sub resource, see:

https://github.com/kubernetes/kubernetes/issues/123541

**Comment by thesuperzapper** [comment]
Commenting to keep this issue alive.

To discuss my current proposal around extending the `Scale` sub resource, see:

https://github.com/kubernetes/kubernetes/issues/123541

/remove-lifecycle rotten

**Comment by thesuperzapper** [comment]
Omg, stale bots are annoying.

/lifecycle frozen

**Comment by sftim** [comment]
Why's this right to be frozen, @thesuperzapper ?

**Comment by alculquicondor** [comment]
Right, if there is a new proposal, it's only fair that we close this one. The history will still be there.

**Comment by sftim** [comment]
/remove-lifecycle frozen
/lifecycle rotten

I suggest also: /close not-planned
with a hint about the new issue (and, even then, formally we recommend the KEP process)
